enhancements/workload-partitioning/wide-availability-workload-partitioning.md

```diff
@@ -0,0 +1,622 @@
+eliminate the race condition that can occur if we apply a machine config after
+the fact as nodes join the cluster. When a customer wishes to pin the management
+workloads they will be able to do that via the existing Performance Profile.
+Resizing partition size will not cause any issues after installation.
+
+In order to implement this enhancement we are proposing changing 4 components
+defined below.
+
+1. Openshift API - ([Infrastructure
+   Status](https://github.com/openshift/api/blob/81fadf1ca0981f800029bc5e2fe2dc7f47ce698b/config/v1/types_infrastructure.go#L51))
+
+   - The change in this component is to store a global identifier if we have
+     partitioning turned on.
+
+2. [Openshift Installer](https://github.com/openshift/installer))
+
+   - The change in the installer is to support turning on this feature only
+     during installation as well as apply the partitioning machine configs so
+     that we avoid the race conditions when running in multi-node environments.
+
+3. Admission Controller ([management cpus
+   override](https://github.com/openshift/kubernetes/blob/a9d6306a701d8fa89a548aa7e132603f6cd89275/openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/doc.go))
+   in openshift/kubernetes.
+   - This change will be in support of checking the global identifier in order
+     to modify the pod spec with the correct `requests`.
+4. The
+   [Performance Profile Controller](https://github.com/openshift/cluster-node-tuning-operator/blob/master/docs/performanceprofile/performance_profile.md)
+   part of [Cluster Node Tuning
+   Operator](https://github.com/openshift/cluster-node-tuning-operator)
+   - This change will support adding the ability to explicitly pin
+     `Infrastructure` workloads.
+
+### Openshift API - Infrastructure Status
+
+We will need to maintain a status flag to be able to identify if a cluster we
+are operating in has been setup for partitioning or not. Since this identifier
+signifies that a cluster's infrastructure is setup for workload partitioning we
+feel that this information should be part of the `Infrastructure Status`
+resource. This identifier will be an enum that will be set during installation
+and during an upgrade for existing single-node deployments we will set it via
+the Node Tuning Operator.
+
+We propose that in the current implementation we only support either `None` or
+`AllNodes`.
+
+```go
+type InfrastructureStatus struct {
+	...
+  // cpuPartitioning expresses if the cluster nodes have CPU Set partitioning turned on.
+  // The default of None means that no CPU Set partitioning is set.
+  // If AllNodes is set that indicates that all the nodes have partitioning set on
+  // and workloads might be pinned to specific CPU Sets depending on the configurations
+  // set via the Node Tuning Operator and the Performance Profile API
+	// +kubebuilder:default=None
+	// +kubebuilder:validation:Enum=None;AllNodes
+	CPUPartitioning PartitioningMode `json:"cpuPartitioning"`
```

Comment:
1. If this resource were to be deleted, how would we re-determine this value by inspecting the cluster?
2. Is there a future where a cluster-admin is allowed to change this value after the cluster is created?

Reply:
1. Currently the plan is to determine this by inspecting the nodes via the NTO, since we'll need to do that for upgrades. We'll also support setting it up via the manifests during install time as well.
2. To my knowledge there is no up coming plan or conversation to allow changing this value in the future, it's all or nothing currently. The thing that holds this back as a toggle feature for now is a timing and scheduling issue. I think if we revisit how we use the `requests` with scheduling, then this feature can be revisited as something that an admin can toggle.